python/packages/zarathustra/skills/subgraph_query_abci/behaviours.py

```diff
@@ -64,40 +68,73 @@ class CheckSubgraphsHealthBehaviour(SubgraphQueryBaseBehaviour):
 
     matching_round: Type[AbstractRound] = CheckSubgraphsHealthRound
 
-    # TODO: implement logic required to set payload content for synchronization
     def async_act(self) -> Generator:
         """Do the act, supporting asynchronous execution."""
 
         with self.context.benchmark_tool.measure(self.behaviour_id).local():
             sender = self.context.agent_address
-            payload = CheckSubgraphsHealthPayload(sender=sender, content="dummy_content")
+            ledger_block_number = 20_000_000  # TODO
+            # ledger_block_number = yield from self.get_latest_block_from_ledger()
+            self.context.logger.info(f"Latest ledger block: {ledger_block_number}")
+            subgraph_block_number = self.get_latest_block_from_subgraph()
+            self.context.logger.info(f"Latest subgraph block: {subgraph_block_number}")
+
+            # as subgraph is often slightly behind, we use a tolerance threshold
+            tolerance = 10
+            subgraph_health = ledger_block_number - subgraph_block_number < tolerance
+            payload = CheckSubgraphsHealthPayload(sender=sender, content=subgraph_health)
 
         with self.context.benchmark_tool.measure(self.behaviour_id).consensus():
             yield from self.send_a2a_transaction(payload)
             yield from self.wait_until_round_end()
 
-        self.set_done()
+    def get_latest_block_from_subgraph(self) -> int:
+        api_key = self.params.config["subgraph_api_key"]
```

Comment:
making a lot of assumpptions

TTHis wont work for self hosted subgraphs.

Check for the api key and see if it can be inserted, potehrwise just the the url.

Reply:
not entirely sure what you mean here. The aea-config requires it to be set

https://github.com/Karrenbelt/verification-station/blob/bd52d9c0583910e3eeaba51b9ae62bf32e4bcacc/python/packages/zarathustra/agents/oracle_verifier/aea-config.yaml#L124-L126

too stringent? What are requirements for self hosted subgraph? Just some (potentially local?) endpoint that requires no token?